### ISSUE TITLE
fail2ban formula depends on python 3.12 but fail2ban 1.0.2 does not work with that python version

### ISSUE THREAD

**Issue by pm215** [Homebrew/homebrew-core#165623]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
lioth$ brew config
HOMEBREW_VERSION: 4.2.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: d7d4c8266210c024b93a450a7d357cec0b46a1bb
Last commit: 5 days ago
Core tap JSON: 09 Mar 16:27 UTC
Core cask tap JSON: 09 Mar 16:27 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.bU5rHfdtvM/org.macosforge.xquartz:0
HOMEBREW_EDITOR: jmacs
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.4-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: 15.3
Rosetta 2: false
lioth$ brew doctor
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

Install a new version of fail2ban

### What happened (include all command output)?

In commit https://github.com/Homebrew/homebrew-core/commit/7956e6ca0789e66f0a1959f273ee60f8e5ee8c5c the fail2ban formula was updated to depend on python 3.12. Unfortunately the current 1.0.2 version of fail2ban does not work with that version of python. Here's the upstream bug about that:

https://github.com/fail2ban/fail2ban/issues/3487

The bug is closed because it has been fixed in upstream's head-of-git, but there is as yet no new version of fail2ban released that will work with python 3.12.

A workaround for this which worked for me was to download the fail2ban.rb from before the erroneous change and install that. This bug could be fixed either by reverting the change in the formula (i.e. stick to python 3.11 for this fail2ban version) or else by backporting the 3.12 fixes from upstream.

This is a repro of the bug, using 'brew reinstall'. Install works fine but trying to actually start the server produces the same error message about asynchat as noted in the upstream bug.

```
lioth$ brew reinstall fail2ban
==> Downloading https://ghcr.io/v2/homebrew/core/fail2ban/manifests/1.0.2-2
Already downloaded: /Users/pm215/Library/Caches/Homebrew/downloads/0310fca8a6f3c3406459204433a100600d924a31c04fd60b411c4d2bde60733a--fail2ban-1.0.2-2.bottle_manifest.json
==> Fetching fail2ban
==> Downloading https://ghcr.io/v2/homebrew/core/fail2ban/blobs/sha256:0b85f9c530ea11f18748d36b4385b9350592b56f8086c74f513372bff3968878
############################################################################################################################################ 100.0%
==> Reinstalling fail2ban 
==> Pouring fail2ban--1.0.2.arm64_sonoma.bottle.2.tar.gz
==> Caveats
Before using Fail2Ban for the first time you should edit the jail
configuration and enable the jails that you want to use, for instance
ssh-ipfw. Also, make sure that they point to the correct configuration
path. I.e. on Mountain Lion the sshd logfile should point to
/var/log/system.log.

  * /opt/homebrew/etc/fail2ban/jail.conf

The Fail2Ban wiki has two pages with instructions for macOS Server that
describes how to set up the Jails for the standard macOS Server
services for the respective releases.

  10.4: https://www.fail2ban.org/wiki/index.php/HOWTO_Mac_OS_X_Server_(10.4)
  10.5: https://www.fail2ban.org/wiki/index.php/HOWTO_Mac_OS_X_Server_(10.5)

Please do not forget to update your configuration files.
They are in /opt/homebrew/etc/fail2ban.

To start fail2ban now and restart at startup:
  sudo brew services start fail2ban
Or, if you don't want/need a background service you can just run:
  /opt/homebrew/opt/fail2ban/bin/fail2ban-client -x start
==> Summary
🍺  /opt/homebrew/Cellar/fail2ban/1.0.2: 345 files, 2.6MB
==> Running `brew cleanup fail2ban`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
lioth$ fail2ban-client -x start
2024-03-09 16:30:49,852 fail2ban.configreader   [6098]: WARNING 'allowipv6' not defined in 'Definition'. Using default one: 'auto'
2024-03-09 16:30:50,088 fail2ban                [6099]: ERROR   No module named 'asynchat'
2024-03-09 16:31:20,621 fail2ban                [6098]: ERROR   Could not start server. Maybe an old socket file is still present. Try to remove /opt/homebrew/var/run/fail2ban/fail2ban.sock. If you used fail2ban-client to start the server, adding the -x option will do it
```

### What did you expect to happen?

The service not to fail with a "No module named 'asynchat'" error.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1: brew install fail2ban
2: fail2ban-client -x start
```

**Comment by SMillerDev** [comment]
Do you have a suggestion for a command to run to catch this next time?

**Comment by pm215** [comment]
I'm not sure. It looks like upstream has some kind of a test suite, so running that would presumably be one option. The other risk-averse approach would be "don't upgrade the Python version to latest-and-greatest unless you've checked with upstream about compatibility", since Python doesn't seem to be very enthusiastic about maintaining back-compat and these things tend to show up as runtime errors rather than buildtime.

**Comment by iMichka** [comment]
Let's fix this first, I opened a pull request: #165924

We will have to wait for a new release as the upstream patches do not seem to apply cleanly.

An improved test (not the full test suite) would be nice though.
